cassiopeia/tools/tree.py

```diff
@@ -0,0 +1,227 @@
+from typing import List, Tuple
+
+import networkx as nx
+
+
+class Tree:
+    r"""
+    A phylogenetic tree for holding data from lineages and lineage tracing
+    experiments.
+
+    (Currently implemented as a light wrapper over networkx.DiGraph)
+
+    Args:
+        tree: The networkx.DiGraph from which to create the tree.
+    """
+
+    def __init__(self, tree: nx.DiGraph):
+        self.tree = tree
+
+    def root(self) -> int:
+        tree = self.tree
+        root = [n for n in tree if tree.in_degree(n) == 0][0]
+        return root
+
+    def leaves(self) -> List[int]:
+        tree = self.tree
+        leaves = [
+            n
+            for n in tree
+            if tree.out_degree(n) == 0 and tree.in_degree(n) == 1
+        ]
+        return leaves
+
+    def internal_nodes(self) -> List[int]:
+        tree = self.tree
+        return [n for n in tree if n != self.root() and n not in self.leaves()]
+
+    def non_root_nodes(self) -> List[int]:
+        return self.leaves() + self.internal_nodes()
+
+    def nodes(self):
+        tree = self.tree
+        return list(tree.nodes())
+
+    def num_characters(self) -> int:
+        return len(self.tree.nodes[self.root()]["characters"])
+
+    def get_state(self, node: int) -> str:
+        tree = self.tree
+        return tree.nodes[node]["characters"]
+
+    def set_state(self, node: int, state: str) -> None:
+        tree = self.tree
+        tree.nodes[node]["characters"] = state
+
+    def set_states(self, node_state_list: List[Tuple[int, str]]) -> None:
+        for (node, state) in node_state_list:
+            self.set_state(node, state)
```

Comment:
Do you want to add a check to make sure that the `node_state_list` argument stores the correct number of characters? It would be bad if one node had 10 characters and the next had 9, right?

Reply:
Yeah that would be a good check. As I mentioned before I think that a bunch of checks could be added in many places but I didn't really focus on that; I was more interested in thinking about the interfaces.